mava/core_jax.py

```diff
@@ -60,18 +60,41 @@ def launch(self) -> None:
         """Run the system."""
 
 
-class SystemBuilder(abc.ABC):
-    """Abstract system builder."""
+class BaseSystemComponent(abc.ABC):
+    """Base class for core components: builder, executor, trainer and dataserver"""
 
-    def __init__(
-        self,
-    ) -> None:
-        """System building init"""
+    def __init__(self) -> None:
+        """Core component init"""
+        self.store: SimpleNamespace = SimpleNamespace()
+        self.callbacks: List
+
+    def has(self, component: Type, subtypes: bool = False) -> bool:
+        """Checks if a component of the given type is in the system.
+
+        Args:
+            component: type of component for which to check.
+            subtypes: whether or not to check for subtypes of the given component.
 
-        # Simple namespace for assigning system builder attributes dynamically
-        self.store = SimpleNamespace()
+        Returns:
+            True if the system has a component of this type.
+        """
+        if subtypes:  # checks for type of component and its subtypes
```

Comment:
Are there instances where we must absolutely specify which test to use? If we don't have such instances then maybe we can try both approaches and remove the argument `subtypes`.

Reply:
Was thinking the same as it's not the cleanest, but I think with some of the DQN stuff we want to be a bit stricter as there's quite a bit of inheritance there when we have DQN > recurrent DQN > VDN > QMIX. So we might want to be sure a specific component is in and no one of it's child classes.

So not sure if it's definitely necessary, but I think it might be in the future.

Reply:
Okay cool. I am happy with it then.